// @jaspero/ng-helpers/esm2015/directives/stop-propagation/stop-propagation.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output, Renderer2 } from '@angular/core';
/**
 * Used for preventing propagation on event calls event.stopPropagation())
 *
 * \@example
 * <div (jpStopPropagation)="doSomething()"></div>
 */
export class StopPropagationDirective {
    /**
     * @param {?} _renderer
     * @param {?} _el
     */
    constructor(_renderer, _el) {
        this._renderer = _renderer;
        this._el = _el;
        /**
         * Any valid html event
         */
        this.stopPropagationEventType = 'click';
        /**
         * Should preventDefault also be called
         */
        this.preventDefault = false;
        /**
         * Outputs the input event
         */
        this.jpStopPropagation = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._renderer.listen(this._el.nativeElement, this.stopPropagationEventType, event => {
            if (this.preventDefault) {
                event.preventDefault();
            }
            if (this.condition !== undefined) {
                if (typeof this.condition === 'boolean') {
                    if (this.condition) {
                        this.sp(event);
                    }
                }
                else if (this.condition(event)) {
                    this.sp(event);
                }
            }
            else {
                this.sp(event);
            }
        });
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    sp(event) {
        event.stopPropagation();
        this.jpStopPropagation.emit(event);
    }
}
StopPropagationDirective.decorators = [
    { type: Directive, args: [{
                selector: '[jpStopPropagation]'
            },] }
];
/** @nocollapse */
StopPropagationDirective.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
StopPropagationDirective.propDecorators = {
    stopPropagationEventType: [{ type: Input }],
    preventDefault: [{ type: Input }],
    condition: [{ type: Input }],
    jpStopPropagation: [{ type: Output }]
};
if (false) {
    /**
     * Any valid html event
     * @type {?}
     */
    StopPropagationDirective.prototype.stopPropagationEventType;
    /**
     * Should preventDefault also be called
     * @type {?}
     */
    StopPropagationDirective.prototype.preventDefault;
    /** @type {?} */
    StopPropagationDirective.prototype.condition;
    /**
     * Outputs the input event
     * @type {?}
     */
    StopPropagationDirective.prototype.jpStopPropagation;
    /**
     * @type {?}
     * @private
     */
    StopPropagationDirective.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    StopPropagationDirective.prototype._el;
}
