// @jaspero/ng-helpers/esm2015/directives/load-click/load-click.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Inject, Input, Renderer2 } from '@angular/core';
import { finalize } from 'rxjs/operators';
import { LOAD_CLICK_CLASS } from './load-click-class.const';
/**
 * Directive will add loading class to the host element on click event
 * Usage: [jpLoadClick]="save()"
 * Function save() should return observable
 */
export class LoadClickDirective {
    /**
     * @param {?} _el
     * @param {?} _renderer
     * @param {?} _defaultLoadClickClass
     */
    constructor(_el, _renderer, _defaultLoadClickClass) {
        this._el = _el;
        this._renderer = _renderer;
        this._defaultLoadClickClass = _defaultLoadClickClass;
        this.loadClickEventType = 'click';
        this.loadClickStopPropagation = false;
        this.loadClickPreventDefault = false;
        this.disableAttribute = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._renderer.listen(this._el.nativeElement, this.loadClickEventType, event => {
            /** @type {?} */
            const defaultClass = this.loadClickClass || this._defaultLoadClickClass;
            if (this.loadClickStopPropagation) {
                event.stopPropagation();
            }
            if (this.loadClickPreventDefault) {
                event.preventDefault();
            }
            this._renderer.addClass(this._el.nativeElement, defaultClass);
            if (this.disableAttribute) {
                this._renderer.setAttribute(this._el.nativeElement, 'disabled', '');
            }
            this.subscription = this.jpLoadClick()
                .pipe(finalize(() => {
                this._renderer.removeClass(this._el.nativeElement, defaultClass);
                if (this.disableAttribute) {
                    this._renderer.removeAttribute(this._el.nativeElement, 'disabled');
                }
            }))
                .subscribe();
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
}
LoadClickDirective.decorators = [
    { type: Directive, args: [{ selector: '[jpLoadClick]' },] }
];
/** @nocollapse */
LoadClickDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: String, decorators: [{ type: Inject, args: [LOAD_CLICK_CLASS,] }] }
];
LoadClickDirective.propDecorators = {
    jpLoadClick: [{ type: Input }],
    loadClickEventType: [{ type: Input }],
    loadClickStopPropagation: [{ type: Input }],
    loadClickPreventDefault: [{ type: Input }],
    loadClickClass: [{ type: Input }],
    disableAttribute: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    LoadClickDirective.prototype.jpLoadClick;
    /** @type {?} */
    LoadClickDirective.prototype.loadClickEventType;
    /** @type {?} */
    LoadClickDirective.prototype.loadClickStopPropagation;
    /** @type {?} */
    LoadClickDirective.prototype.loadClickPreventDefault;
    /** @type {?} */
    LoadClickDirective.prototype.loadClickClass;
    /** @type {?} */
    LoadClickDirective.prototype.disableAttribute;
    /** @type {?} */
    LoadClickDirective.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    LoadClickDirective.prototype._el;
    /**
     * @type {?}
     * @private
     */
    LoadClickDirective.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    LoadClickDirective.prototype._defaultLoadClickClass;
}
