// @jaspero/ng-helpers/esm2015/directives/load-click/load-click.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { LoadClickDirective } from './load-click.directive';
import { LOAD_CLICK_CLASS } from './load-click-class.const';
export class LoadClickModule {
    /**
     * @param {?} value
     * @return {?}
     */
    static defaultLoadingClass(value) {
        return {
            ngModule: LoadClickModule,
            providers: [
                {
                    provide: LOAD_CLICK_CLASS,
                    useValue: value
                }
            ]
        };
    }
}
LoadClickModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    LoadClickDirective
                ],
                exports: [
                    LoadClickDirective
                ],
                providers: [
                    {
                        provide: LOAD_CLICK_CLASS,
                        useValue: 'loading'
                    }
                ]
            },] }
];
