// @jaspero/form-builder/esm2015/lib/fields/slider/slider.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FieldComponent } from '../../field/field.component';
export class SliderComponent extends FieldComponent {
    ngOnInit() {
        this.startAt = this.cData.starAt
            ? this.cData.starAt
            : this.cData.validation.minimum
                ? this.cData.validation.minimum
                : 0;
        this.endAt = this.cData.endAt
            ? this.cData.endAt
            : this.cData.validation.maximum
                ? this.cData.validation.maximum
                : 100;
    }
}
SliderComponent.decorators = [
    { type: Component, args: [{
                selector: 'fb-slider',
                template: "<div class=\"mat-form-field-wrapper\">\n  <p class=\"fs-small fg-secondary\" *ngIf=\"cData.label\">{{(cData.label || '') | transloco}}</p>\n  <mat-slider\n    [formControl]=\"cData.control\"\n    [thumbLabel]=\"cData.thumbLabel\"\n    [tickInterval]=\"cData.tickInterval\"\n    [min]=\"startAt\"\n    [max]=\"endAt\">\n  </mat-slider>\n  <p class=\"fs-small fg-secondary\" *ngIf=\"cData.hint\">\n    {{(cData.hint || '') | transloco}}\n  </p>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
