// @jaspero/form-builder/esm2015/lib/segments/tabs/tabs.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { SegmentComponent } from '../../segment/segment.component';
import { filterAndCompileSegments } from '../../utils/filter-and-compile-segments';
import { safeEval } from '../../utils/safe-eval';
export class TabsComponent extends SegmentComponent {
    get configuration() {
        return this.segment.configuration;
    }
    ngOnInit() {
        super.ngOnInit();
        if (this.configuration.selectedTabChange) {
            this.selectedTabChange = safeEval(this.configuration.selectedTabChange);
        }
        this.tabs = this.configuration.tabs.map(tab => (Object.assign(Object.assign({}, tab), { fields: (tab.fields || []).map(key => this.sData.parser.field(key, this.pointers[key], this.sData.definitions)), nestedSegments: filterAndCompileSegments(tab.nestedSegments || [], this.sData.parser, this.sData.definitions, this.injector, this.segment.entryValue) })));
    }
    tabChange(event) {
        if (this.selectedTabChange) {
            this.selectedTabChange(event, this.sData);
        }
    }
}
TabsComponent.decorators = [
    { type: Component, args: [{
                selector: 'fb-tabs',
                template: "<ng-container *ngIf=\"segment.title || segment.subTitle\">\n  <h4 *ngIf=\"segment.title\">{{segment.title}}</h4>\n  <h5 *ngIf=\"segment.subTitle\">{{segment.subTitle}}</h5>\n  <div class=\"p-y-m\">\n    <mat-divider></mat-divider>\n  </div>\n</ng-container>\n\n<p *ngIf=\"segment.description\">{{segment.description}}</p>\n\n<mat-tab-group\n  [selectedIndex]=\"configuration.selectedIndex\"\n  [dynamicHeight]=\"configuration.dynamicHeight\"\n  [disableRipple]=\"configuration.disableRipple\"\n  [attr.mat-align-tabs]=\"configuration.alignment\"\n  (selectedTabChange)=\"tabChange($event)\">\n\n  <mat-tab\n    *ngFor=\"let tab of tabs\"\n    [label]=\"tab.title\"\n    [disabled]=\"tab.disabled\">\n\n    <div class=\"p-a-s\">\n      <ng-container *ngFor=\"let field of tab.fields | showField:sData.parser\">\n        <ng-template [cdkPortalOutlet]=\"field.portal\"></ng-template>\n      </ng-container>\n    </div>\n\n    <ng-container *ngFor=\"let nested of tab.nestedSegments\">\n      <ng-template [cdkPortalOutlet]=\"nested.component\"></ng-template>\n    </ng-container>\n  </mat-tab>\n</mat-tab-group>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
