// @jaspero/form-builder/esm2015/lib/fields/wysiwyg/wysiwyg.component.js
import { ChangeDetectionStrategy, Component, Inject, ViewChild } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
import { MatDialog } from '@angular/material/dialog';
import { filter, take } from 'rxjs/operators';
import { FieldComponent } from '../../field/field.component';
import 'tinymce/plugins/code';
import 'tinymce/plugins/print';
import 'tinymce/plugins/wordcount';
import 'tinymce/plugins/link';
import 'tinymce/plugins/image';
import 'tinymce/plugins/imagetools';
import 'tinymce/plugins/fullscreen';
import 'tinymce/plugins/lists';
import 'tinymce/plugins/autolink';
import 'tinymce/plugins/advlist';
import 'tinymce/plugins/table';
import { FormBuilderComponent } from '../../form-builder.component';
import { StorageService } from '../../services/storage.service';
import { COMPONENT_DATA } from '../../utils/create-component-injector';
export class WysiwygComponent extends FieldComponent {
    constructor(cData, fb, dialog, storage, formBuilderComponent) {
        super(cData);
        this.cData = cData;
        this.fb = fb;
        this.dialog = dialog;
        this.storage = storage;
        this.formBuilderComponent = formBuilderComponent;
        this.ytDefault = {
            fullWidth: true,
            showPlayerControls: true,
            privacyEnhancedMode: false,
            align: 'left'
        };
    }
    ngOnInit() {
        this.ytForm = this.fb.group(Object.assign({ value: ['', Validators.required] }, this.ytDefault));
        this.cData.control.statusChanges.subscribe(value => {
            if (value === 'DISABLED') {
                tinymce.activeEditor.getBody().setAttribute('readonly', true);
            }
            else if (this.cData.control.disabled) {
                tinymce.activeEditor.getBody().setAttribute('readonly', false);
            }
        });
    }
    ngAfterViewInit() {
        this.registerTiny();
    }
    registerTiny() {
        tinymce.init({
            target: this.textarea.nativeElement,
            branding: false,
            height: this.cData.height || 420,
            plugins: [
                'code',
                'print',
                'wordcount',
                'link',
                'lists',
                'advlist',
                'autolink',
                'image',
                'imagetools',
                'fullscreen',
                'table'
            ],
            menubar: this.cData.menubar || 'edit insert view format table tools help',
            image_advtab: true,
            /**
             * Link settings
             */
            default_link_target: '_blank',
            readonly: this.cData.control.disabled,
            toolbar: this.cData.toolbar || [
                'undo redo',
                'insert',
                'styleselect',
                'bold italic',
                'forecolor backcolor',
                'alignleft aligncenter alignright alignjustify',
                'bullist numlist outdent indent',
                'link',
                'image',
                'youTube',
                'fullscreen',
            ].join(' | '),
            images_upload_handler: (blobInfo, success, failure) => {
                this.storage
                    .upload(blobInfo.filename(), blobInfo.blob(), this.formBuilderComponent.id ? {
                    customMetadata: {
                        collection: this.formBuilderComponent.id
                    }
                } : {})
                    .then((data) => data.ref.getDownloadURL())
                    .then((url) => success(url))
                    .catch((error) => failure(error.toString()));
            },
            setup: (editor) => {
                this.editor = editor;
                editor.on('keyup change', () => {
                    const tinyContent = editor.getContent();
                    this.cData.control.setValue(tinyContent);
                });
                editor.ui.registry.addButton('youTube', {
                    type: 'button',
                    icon: 'embed',
                    tooltip: 'Embed youtube video',
                    onAction: () => {
                        this.ytForm.reset(this.ytDefault);
                        this.dialog
                            .open(this.youTubeDialogTemplate, { width: '500px' })
                            .afterClosed()
                            .pipe(filter(value => !!value), take(1))
                            .subscribe(() => {
                            const data = this.ytForm.getRawValue();
                            let url = data.privacyEnhancedMode
                                ? 'https://www.youtube-nocookie.com/embed/'
                                : 'https://www.youtube.com/embed/';
                            url = url + data.value;
                            if (!data.showPlayerControls) {
                                data.value += '?controls=0';
                            }
                            const iframe = `<iframe width="560" height="315" src="${url}" frameborder="0"
                    allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>`;
                            editor.insertContent(data.fullWidth
                                ? `<div class="evw-full" style="text-align: ${data.align}">${iframe}</div>`
                                : `<div style="text-align: ${data.align}">${iframe}</div>`);
                        });
                    }
                });
            }
        });
    }
}
WysiwygComponent.decorators = [
    { type: Component, args: [{
                selector: 'fb-wysiwyg',
                template: "<div class=\"mat-form-field-wrapper\">\n  <p class=\"fs-small fg-secondary\" *ngIf=\"cData.label\">{{(cData.label || '') | transloco}}</p>\n  <textarea #textarea role=\"textarea\">{{cData.control.value}}</textarea>\n</div>\n\n<ng-template #youTubeDialog>\n  <h1 mat-dialog-title>{{'FIELDS.WYSIWYG.EMBED_YOUTUBE' | transloco}}</h1>\n\n  <form [formGroup]=\"ytForm\">\n    <mat-dialog-content>\n      <mat-form-field class=\"w-full\">\n        <mat-label>{{'FIELDS.WYSIWYG.VIDEO_ID' | transloco}}</mat-label>\n        <input matInput formControlName=\"value\">\n      </mat-form-field>\n\n      <label class=\"mat-form-field-wrapper flex jc-between\">\n        <span>{{'FIELDS.WYSIWYG.FULL_WIDTH' | transloco}}</span>\n        <mat-slide-toggle formControlName=\"fullWidth\"></mat-slide-toggle>\n      </label>\n\n      <label class=\"mat-form-field-wrapper flex jc-between\">\n        <span>{{'FIELDS.WYSIWYG.SHOW_PLAYER_CONTROLS' | transloco}}</span>\n        <mat-slide-toggle formControlName=\"showPlayerControls\"></mat-slide-toggle>\n      </label>\n\n      <label class=\"mat-form-field-wrapper flex jc-between\">\n        <span>{{'FIELDS.WYSIWYG.PRIVACY_ENHANCED_MODE' | transloco}}</span>\n        <mat-slide-toggle formControlName=\"privacyEnhancedMode\"></mat-slide-toggle>\n      </label>\n\n      <mat-form-field class=\"w-full\">\n        <mat-label>{{'FIELDS.WYSIWYG.ALIGNMENT' | transloco}}</mat-label>\n        <mat-select formControlName=\"align\">\n          <mat-option value=\"left\">{{'FIELDS.WYSIWYG.LEFT' | transloco}}</mat-option>\n          <mat-option value=\"center\">{{'FIELDS.WYSIWYG.CENTER' | transloco}}</mat-option>\n          <mat-option value=\"right\">{{'FIELDS.WYSIWYG.RIGHT' | transloco}}</mat-option>\n        </mat-select>\n      </mat-form-field>\n    </mat-dialog-content>\n\n    <mat-dialog-actions class=\"jc-end\">\n      <button mat-button mat-dialog-close>{{'GENERAL.CLOSE' | transloco}}</button>\n      <button mat-raised-button color=\"primary\" [mat-dialog-close]=\"true\">{{'GENERAL.SAVE' | transloco}}</button>\n    </mat-dialog-actions>\n  </form>\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
WysiwygComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [COMPONENT_DATA,] }] },
    { type: FormBuilder },
    { type: MatDialog },
    { type: StorageService },
    { type: FormBuilderComponent }
];
WysiwygComponent.propDecorators = {
    textarea: [{ type: ViewChild, args: ['textarea', { static: true },] }],
    youTubeDialogTemplate: [{ type: ViewChild, args: ['youTubeDialog', { static: true },] }]
};
