// @jaspero/ng-helpers/esm2015/pipes/sanitize/sanitize.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
/**
 * Applies the appropriate DomSanitizer method
 * to inputted value.
 *
 * \@example
 * <div [innerHtml]="someHtmlValue | sanitize"></div>
 */
export class SanitizePipe {
    /**
     * @param {?} _sanitizer
     */
    constructor(_sanitizer) {
        this._sanitizer = _sanitizer;
    }
    /**
     * @param {?} value
     * @param {?=} type
     * @return {?}
     */
    transform(value, type = 'html') {
        /** @type {?} */
        const sanitizeMap = {
            html: 'bypassSecurityTrustHtml',
            style: 'bypassSecurityTrustStyle',
            script: 'bypassSecurityTrustScript',
            url: 'bypassSecurityTrustUrl',
            resourceUrl: 'bypassSecurityTrustResourceUrl'
        };
        return this._sanitizer[sanitizeMap[type]](value);
    }
}
SanitizePipe.decorators = [
    { type: Pipe, args: [{
                name: 'jpSanitize'
            },] }
];
/** @nocollapse */
SanitizePipe.ctorParameters = () => [
    { type: DomSanitizer }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    SanitizePipe.prototype._sanitizer;
}
