// @jaspero/ng-helpers/esm2015/pipes/time-passed/time-passed.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
/** @enum {number} */
const TimePassedType = {
    Millisecond: 0,
    Second: 1,
    Minute: 2,
    Hour: 3,
    Day: 4,
    Month: 5,
    Year: 6,
};
export { TimePassedType };
TimePassedType[TimePassedType.Millisecond] = 'Millisecond';
TimePassedType[TimePassedType.Second] = 'Second';
TimePassedType[TimePassedType.Minute] = 'Minute';
TimePassedType[TimePassedType.Hour] = 'Hour';
TimePassedType[TimePassedType.Day] = 'Day';
TimePassedType[TimePassedType.Month] = 'Month';
TimePassedType[TimePassedType.Year] = 'Year';
export class TimePassedPipe {
    /**
     * @param {?} dateOne
     * @param {?=} dateTwo
     * @param {?=} type
     * @return {?}
     */
    static timeDiff(dateOne, dateTwo = new Date(), type = TimePassedType.Day) {
        /** @type {?} */
        const oneDay = 24 * 60 * 60 * 1000;
        /** @type {?} */
        const minute = 60 * 1000;
        /** @type {?} */
        const hour = 60 * minute;
        /** @type {?} */
        const day = hour * 24;
        /** @type {?} */
        const durationMap = {
            [TimePassedType.Millisecond]: 1,
            [TimePassedType.Second]: 1000,
            [TimePassedType.Minute]: minute,
            [TimePassedType.Hour]: hour,
            [TimePassedType.Day]: day,
            [TimePassedType.Month]: day * 30,
            [TimePassedType.Year]: day * 365
        };
        return Math.round(Math.abs((dateOne.getTime() - dateTwo.getTime()) / durationMap[type]));
    }
    /**
     * @param {?} dateOne
     * @param {?=} dateTwo
     * @param {?=} type
     * @return {?}
     */
    transform(dateOne, dateTwo, type) {
        if (!dateTwo) {
            dateTwo = new Date();
        }
        return TimePassedPipe.timeDiff(dateOne, dateTwo, type !== undefined ? type : TimePassedType.Minute);
    }
}
TimePassedPipe.decorators = [
    { type: Pipe, args: [{
                name: 'jpTimePassed'
            },] }
];
