// @jaspero/form-builder/esm2015/lib/directives/dropzone/dropzone.directive.js
import { Directive, ElementRef, EventEmitter, HostListener, Input, Output, Renderer2 } from '@angular/core';
export class DropzoneDirective {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.hoverClass = 'active';
        this.dropped = new EventEmitter();
        this.hovered = new EventEmitter();
    }
    onDrop($event) {
        $event.preventDefault();
        this.dropped.emit($event.dataTransfer.files);
        this.renderer.removeClass(this.el.nativeElement, this.hoverClass);
        this.hovered.emit(false);
    }
    onDragOver($event) {
        $event.preventDefault();
        this.renderer.addClass(this.el.nativeElement, this.hoverClass);
        this.hovered.emit(true);
    }
    onDragLeave($event) {
        $event.preventDefault();
        this.renderer.removeClass(this.el.nativeElement, this.hoverClass);
        this.hovered.emit(false);
    }
}
DropzoneDirective.decorators = [
    { type: Directive, args: [{
                selector: '[fbDropzone]'
            },] }
];
DropzoneDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
DropzoneDirective.propDecorators = {
    hoverClass: [{ type: Input }],
    dropped: [{ type: Output }],
    hovered: [{ type: Output }],
    onDrop: [{ type: HostListener, args: ['drop', ['$event'],] }],
    onDragOver: [{ type: HostListener, args: ['dragover', ['$event'],] }],
    onDragLeave: [{ type: HostListener, args: ['dragleave', ['$event'],] }]
};
