// @jaspero/ng-helpers/esm2015/helpers/on-change.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Courtesy of
 * https://blog.angularindepth.com/creatively-decouple-ngonchanges-fab95395cc6e
 */
/**
 * @record
 * @template T
 */
export function SimpleChange() { }
if (false) {
    /** @type {?} */
    SimpleChange.prototype.firstChange;
    /** @type {?} */
    SimpleChange.prototype.previousValue;
    /** @type {?} */
    SimpleChange.prototype.currentValue;
    /** @type {?} */
    SimpleChange.prototype.isFirstChange;
}
/**
 * @template T
 * @param {?} callback
 * @return {?}
 */
export function OnChange(callback) {
    /** @type {?} */
    const cachedValueKey = Symbol();
    /** @type {?} */
    const isFirstChangeKey = Symbol();
    return (target, key) => {
        Object.defineProperty(target, key, {
            set: function (value) {
                /**
                 * Change status of "isFirstChange"
                 */
                this[isFirstChangeKey] = this[isFirstChangeKey] === undefined;
                /**
                 *  No operation if new value is same as old value
                 */
                if (!this[isFirstChangeKey] && this[cachedValueKey] === value) {
                    return;
                }
                /** @type {?} */
                const oldValue = this[cachedValueKey];
                this[cachedValueKey] = value;
                /** @type {?} */
                const simpleChange = {
                    firstChange: this[isFirstChangeKey],
                    previousValue: oldValue,
                    currentValue: this[cachedValueKey],
                    isFirstChange: () => this[isFirstChangeKey],
                };
                callback.call(this, this[cachedValueKey], simpleChange);
            },
            get: function () {
                return this[cachedValueKey];
            }
        });
    };
}
