// @jaspero/form-builder/esm2015/lib/fields/date-field/date-field.component.js
import { DatePipe } from '@angular/common';
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Inject } from '@angular/core';
import { FormControl, Validators } from '@angular/forms';
import { of } from 'rxjs';
import { tap } from 'rxjs/operators';
import { FieldComponent } from '../../field/field.component';
import { FormBuilderService } from '../../form-builder.service';
import { COMPONENT_DATA } from '../../utils/create-component-injector';
import { cloneAbstractControl } from '../../utils/clone-abstract-control';
export class DateFieldComponent extends FieldComponent {
    constructor(cData, cdr, formBuilderService) {
        super(cData);
        this.cData = cData;
        this.cdr = cdr;
        this.formBuilderService = formBuilderService;
    }
    ngOnInit() {
        this.formBuilderService.saveComponents.push(this);
        this.startDate = this.cData.startAt
            ? new Date(this.cData.startAt)
            : new Date();
        const date = new Date(this.cData.control.value);
        this.entryControl = cloneAbstractControl(this.cData.control);
        this.hoursControl = new FormControl({ value: date.getHours() || 0, disabled: this.cData.control.disabled }, [Validators.min(0), Validators.max(23)]);
        this.minutesControl = new FormControl({ value: date.getMinutes() || 0, disabled: this.cData.control.disabled }, [Validators.min(0), Validators.max(59)]);
        /**
         * Dirty hack for getting numbers to display properly might need revisiting
         */
        if (typeof this.entryControl.value === 'number' ||
            typeof this.entryControl.value === 'string') {
            this.entryControl.setValue(new Date(this.entryControl.value));
        }
    }
    save() {
        return of({}).pipe(tap(() => {
            let value = this.entryControl.value;
            if (value) {
                if (this.cData.includeTime) {
                    value.setHours(this.hoursControl.value || 0);
                    value.setMinutes(this.minutesControl.value || 0);
                }
                if (this.cData.format) {
                    if (this.cData.format === 'number') {
                        value = value.getTime();
                    }
                    else {
                        const pipe = new DatePipe('en');
                        value = pipe.transform(value, this.cData.format);
                    }
                }
            }
            this.cData.control.setValue(value);
        }));
    }
}
DateFieldComponent.decorators = [
    { type: Component, args: [{
                selector: 'fb-date-field',
                template: "<div class=\"grid\">\n  <div class=\"p-l-0 {{!cData.includeTime ? 'col-12' : 'col-8'}}\">\n    <mat-form-field class=\"w-full\" appearance=\"outline\">\n      <mat-label>\n        {{(cData.label || '') | transloco}}\n      </mat-label>\n      <input\n        matInput\n        [attr.autocomplete]=\"cData.autocomplete || 'off'\"\n        [matDatepicker]=\"picker\"\n        [placeholder]=\"(cData.placeholder || '') | transloco\"\n        [formControl]=\"entryControl\">\n      <mat-datepicker-toggle matSuffix [for]=\"picker\"></mat-datepicker-toggle>\n      <mat-datepicker\n        #picker\n        [touchUi]=\"cData.touchUi || false\"\n        [startView]=\"cData.startView || 'month'\"\n        [startAt]=\"startDate\">\n      </mat-datepicker>\n      <mat-hint *ngIf=\"cData.hint\" [innerHTML]=\"cData.hint | jpSanitize\"></mat-hint>\n    </mat-form-field>\n  </div>\n  <ng-container *ngIf=\"cData.includeTime\">\n    <div class=\"col-2\">\n      <mat-form-field class=\"w-full\" class=\"w-full\" appearance=\"outline\">\n        <mat-label>{{(cData.labelHours || 'HOURS') | transloco}}</mat-label>\n        <input matInput type=\"number\" [placeholder]=\"cData.placeholderHours || 'HH'\" [formControl]=\"hoursControl\">\n      </mat-form-field>\n    </div>\n    <div class=\"col-2 p-r-0\">\n      <mat-form-field class=\"w-full\" appearance=\"outline\">\n        <mat-label>{{(cData.labelMinutes || 'MINUTES') | transloco}}</mat-label>\n        <input matInput type=\"number\" [placeholder]=\"cData.placeholderMinutes || 'MM'\" [formControl]=\"minutesControl\">\n      </mat-form-field>\n    </div>\n  </ng-container>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
DateFieldComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [COMPONENT_DATA,] }] },
    { type: ChangeDetectorRef },
    { type: FormBuilderService }
];
