// @jaspero/form-builder/esm2015/lib/fields/chips/chips.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FieldComponent } from '../../field/field.component';
import { getHsd } from '../../utils/get-hsd';
export class ChipsComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.data = [];
    }
    ngOnInit() {
        this.data = this.cData.control.value;
        this.removable = this.cData.hasOwnProperty('removable') ? this.cData.removable : true;
        this.prefix$ = getHsd('prefix', this.cData);
        this.suffix$ = getHsd('suffix', this.cData);
    }
    add(event) {
        const value = (event.value || '').trim();
        const input = event.input;
        if (value && (this.cData.unique ? !this.data.includes(value) : true)) {
            this.data.push(value);
            this.cData.control.setValue(this.data);
        }
        if (input) {
            input.value = '';
        }
    }
    remove(chip) {
        const index = this.cData.control.value.indexOf(chip);
        if (index >= 0) {
            this.data.splice(index, 1);
            this.cData.control.setValue(this.data);
        }
    }
}
ChipsComponent.decorators = [
    { type: Component, args: [{
                selector: 'fb-chips',
                template: "<mat-form-field class=\"w-full\" appearance=\"outline\">\n  <mat-label>{{(cData.label || '') | transloco}}</mat-label>\n  <mat-chip-list #chipList>\n    <mat-chip\n      *ngFor=\"let chip of data\"\n      [selectable]=\"cData.selectable\"\n      [removable]=\"removable\"\n      (removed)=\"remove(chip)\">\n      {{chip}}\n      <mat-icon *ngIf=\"removable\" matChipRemove>cancel</mat-icon>\n    </mat-chip>\n    <input\n      matInput\n      [attr.autocomplete]=\"cData.autocomplete || 'on'\"\n      [placeholder]=\"(cData.placeholder || '') | transloco\"\n      (matChipInputTokenEnd)=\"add($event)\"\n      [matChipInputFor]=\"chipList\">\n  </mat-chip-list>\n  <span *ngIf=\"cData.suffix\" matSuffix [innerHTML]=\"((suffix$ | async) || '') | jpSanitize\"></span>\n  <span *ngIf=\"cData.prefix\" matPrefix [innerHTML]=\"((prefix$ | async) || '') | jpSanitize\"></span>\n  <mat-hint *ngIf=\"cData.hint\" [innerHTML]=\"cData.hint | jpSanitize\"></mat-hint>\n</mat-form-field>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
