// @jaspero/form-builder/esm2015/lib/fields/range/range.component.js
import { ChangeDetectorRef, Component, Inject } from '@angular/core';
import { FormControl, Validators } from '@angular/forms';
import { of } from 'rxjs';
import { tap } from 'rxjs/operators';
import { FieldComponent } from '../../field/field.component';
import { FormBuilderService } from '../../form-builder.service';
import { cloneAbstractControl } from '../../utils/clone-abstract-control';
import { COMPONENT_DATA } from '../../utils/create-component-injector';
export class RangeComponent extends FieldComponent {
    constructor(cData, cdr, formBuilderService) {
        super(cData);
        this.cData = cData;
        this.cdr = cdr;
        this.formBuilderService = formBuilderService;
    }
    ngOnInit() {
        this.formBuilderService.saveComponents.push(this);
        this.min = new Date(this.cData.min || 0);
        this.max = new Date(this.cData.max || Date.now() * 2);
        this.entryControl = cloneAbstractControl(this.cData.control);
        this.start = new FormControl({ value: this.entryControl.value.start || '', disabled: this.cData.control.disabled }, [Validators.min(this.cData.min || 0)]);
        this.end = new FormControl({ value: this.entryControl.value.end || '', disabled: this.cData.control.disabled }, [Validators.max(this.cData.max || Date.now() * 2)]);
    }
    save() {
        return of({}).pipe(tap(() => {
            const start = this.start.value;
            const end = this.end.value;
            this.cData.control.setValue({ start, end });
        }));
    }
}
RangeComponent.decorators = [
    { type: Component, args: [{
                selector: 'fb-range',
                template: "<div class=\"mat-form-field-wrapper\">\n  <p class=\"fs-small fg-secondary\" *ngIf=\"cData.label\">{{(cData.label || '') | transloco}}</p>\n\n  <mat-form-field appearance=\"outline\" class=\"w-full\">\n    <mat-label>Enter a date range</mat-label>\n    <mat-date-range-input [rangePicker]=\"picker\">\n      <input matStartDate placeholder=\"Start date\" [formControl]=\"start\" [min]=\"min\">\n      <input matEndDate placeholder=\"End date\" [formControl]=\"end\" [max]=\"max\">\n    </mat-date-range-input>\n    <mat-datepicker-toggle matSuffix [for]=\"picker\"></mat-datepicker-toggle>\n    <mat-date-range-picker #picker></mat-date-range-picker>\n  </mat-form-field>\n\n  <p class=\"fs-small fg-secondary\" *ngIf=\"cData.hint\">\n    {{(cData.hint | transloco) | transloco}}\n  </p>\n</div>\n",
                styles: [""]
            },] }
];
RangeComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [COMPONENT_DATA,] }] },
    { type: ChangeDetectorRef },
    { type: FormBuilderService }
];
