// @jaspero/form-builder/esm2015/lib/fields/autocomplete/autocomplete.component.js
import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { BehaviorSubject, of } from 'rxjs';
import { map, startWith, switchMap, tap } from 'rxjs/operators';
import { DbService } from '../../services/db.service';
import { FieldComponent } from '../../field/field.component';
import { COMPONENT_DATA } from '../../utils/create-component-injector';
import { getHsd } from '../../utils/get-hsd';
export class AutocompleteComponent extends FieldComponent {
    constructor(cData, dbService) {
        super(cData);
        this.cData = cData;
        this.dbService = dbService;
        this.loading$ = new BehaviorSubject(true);
    }
    ngOnInit() {
        this.prefix$ = getHsd('prefix', this.cData);
        this.suffix$ = getHsd('suffix', this.cData);
        let dataSet$;
        if (this.cData.populate) {
            const { populate } = this.cData;
            dataSet$ = this.dbService
                .getDocumentsSimple(this.cData.populate.collection, this.cData.populate.orderBy, this.cData.populate.filter)
                .pipe(map(docs => docs.map(doc => ({
                value: doc[populate.valueKey || 'id'],
                name: doc[populate.nameKey]
            }))), tap(() => this.loading$.next(false)));
        }
        else {
            dataSet$ = of(this.cData.dataSet);
        }
        this.filteredSet$ = dataSet$.pipe(switchMap(dataSet => this.cData.control.valueChanges.pipe(startWith(this.cData.control.value), map(value => {
            if (!value) {
                return dataSet;
            }
            value = value.toLowerCase();
            return dataSet.filter(item => item.name.toLowerCase().includes(value));
        }))));
    }
}
AutocompleteComponent.decorators = [
    { type: Component, args: [{
                selector: 'fb-autocomplete',
                template: "<mat-form-field appearance=\"outline\" class=\"w-full\">\n  <mat-label>{{(cData.label || '') | transloco}}</mat-label>\n  <input\n    matInput\n    [placeholder]=\"(cData.placeholder || '') | transloco\"\n    [attr.autocomplete]=\"cData.autocomplete || 'on'\"\n    [formControl]=\"cData.control\"\n    [matAutocomplete]=\"auto\">\n  <mat-autocomplete #auto=\"matAutocomplete\">\n    <mat-option *ngFor=\"let data of filteredSet$ | async\" [value]=\"data.value\">\n      {{data.name}}\n    </mat-option>\n  </mat-autocomplete>\n  <mat-hint *ngIf=\"cData.hint\" [innerHTML]=\"cData.hint | jpSanitize\"></mat-hint>\n  <span *ngIf=\"cData.suffix\" matSuffix [innerHTML]=\"((suffix$ | async) || '') | jpSanitize\"></span>\n  <span *ngIf=\"cData.prefix\" matPrefix [innerHTML]=\"((prefix$ | async) || '') | jpSanitize\"></span>\n</mat-form-field>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
AutocompleteComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [COMPONENT_DATA,] }] },
    { type: DbService }
];
