// @ngneat/transloco/esm5/lib/transloco-testing.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/transloco-testing.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Inject, Injectable, NgModule } from '@angular/core';
import { TRANSLOCO_LOADER } from './transloco.loader';
import { of } from 'rxjs';
import { defaultProviders, TranslocoModule } from './transloco.module';
import { TRANSLOCO_CONFIG, translocoConfig } from './transloco.config';
var TestingLoader = /** @class */ (function () {
    function TestingLoader(langs) {
        this.langs = langs;
    }
    /**
     * @param {?} lang
     * @return {?}
     */
    TestingLoader.prototype.getTranslation = /**
     * @param {?} lang
     * @return {?}
     */
    function (lang) {
        return of(this.langs[lang]);
    };
    TestingLoader.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    TestingLoader.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: ['translocoLangs',] }] }
    ]; };
    return TestingLoader;
}());
export { TestingLoader };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TestingLoader.prototype.langs;
}
var TranslocoTestingModule = /** @class */ (function () {
    function TranslocoTestingModule() {
    }
    /**
     * @param {?} langs
     * @param {?=} config
     * @return {?}
     */
    TranslocoTestingModule.withLangs = /**
     * @param {?} langs
     * @param {?=} config
     * @return {?}
     */
    function (langs, config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: TranslocoTestingModule,
            providers: [
                {
                    provide: 'translocoLangs',
                    useValue: langs
                },
                {
                    provide: TRANSLOCO_LOADER,
                    useClass: TestingLoader
                },
                defaultProviders,
                {
                    provide: TRANSLOCO_CONFIG,
                    useValue: translocoConfig(tslib_1.__assign({ prodMode: true, missingHandler: { logMissingKey: false } }, config))
                }
            ]
        };
    };
    TranslocoTestingModule.decorators = [
        { type: NgModule, args: [{
                    exports: [TranslocoModule]
                },] }
    ];
    return TranslocoTestingModule;
}());
export { TranslocoTestingModule };
