// @jaspero/form-builder/esm2015/lib/form-builder.component.js
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Inject, Injector, Input, Optional, Output } from '@angular/core';
import { forkJoin, of } from 'rxjs';
import { SegmentType } from './enums/segment-type.enum';
import { State } from './enums/state.enum';
import { FormBuilderService } from './form-builder.service';
import { CUSTOM_FIELDS } from './utils/custom-fields';
import { filterAndCompileSegments } from './utils/filter-and-compile-segments';
import { Parser } from './utils/parser';
import { ROLE } from './utils/role';
export class FormBuilderComponent {
    constructor(injector, role, customFields, cdr, service) {
        this.injector = injector;
        this.role = role;
        this.customFields = customFields;
        this.cdr = cdr;
        this.service = service;
        this.valueChanges = new EventEmitter();
        this.validityChanges = new EventEmitter();
        this.state = State.Create;
    }
    ngOnChanges(changes) {
        if (changes.data) {
            this.render();
        }
        if (changes.value && this.form) {
            this.form.patchValue(changes.value.currentValue);
        }
    }
    ngOnDestroy() {
        if (this.changeSubscription) {
            this.changeSubscription.unsubscribe();
        }
        if (this.statusSubscription) {
            this.statusSubscription.unsubscribe();
        }
    }
    process() {
        this.parser.preSaveHooks(this.state);
        return this.form.getRawValue();
    }
    save(collectionId, documentId) {
        const toExec = this.service.saveComponents.map(comp => comp.save(collectionId, documentId));
        return toExec.length ? forkJoin(toExec) : of({});
    }
    saveAndProcess(collectionId, documentId) {
        this.process();
        return this.save(collectionId, documentId);
    }
    render() {
        const value = this.data.value || {};
        const definitions = this.data.definitions || {};
        this.parser = new Parser(this.data.schema, this.injector, this.state, this.role, definitions, this.customFields);
        this.form = this.parser.buildForm(value, [], '/', false);
        this.parser.loadHooks();
        this.segments = filterAndCompileSegments(this.data.segments ||
            [{
                    title: '',
                    fields: Object.keys(this.parser.pointers),
                    columnsDesktop: 12,
                    type: SegmentType.Empty
                }], this.parser, definitions, this.injector, value);
        if (this.changeSubscription) {
            this.changeSubscription.unsubscribe();
        }
        if (this.statusSubscription) {
            this.statusSubscription.unsubscribe();
        }
        this.changeSubscription = this.form.valueChanges
            .subscribe(val => {
            this.valueChanges.emit(val);
        });
        this.statusSubscription = this.form.statusChanges
            .subscribe(val => {
            this.validityChanges.emit(val === 'VALID');
        });
        this.cdr.markForCheck();
    }
}
FormBuilderComponent.decorators = [
    { type: Component, args: [{
                selector: 'fb-form-builder',
                template: "<div class=\"grid\">\n  <ng-container *ngFor=\"let segment of segments\">\n    <ng-template [cdkPortalOutlet]=\"segment.component\"></ng-template>\n  </ng-container>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [FormBuilderService]
            },] }
];
FormBuilderComponent.ctorParameters = () => [
    { type: Injector },
    { type: String, decorators: [{ type: Optional }, { type: Inject, args: [ROLE,] }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [CUSTOM_FIELDS,] }] },
    { type: ChangeDetectorRef },
    { type: FormBuilderService }
];
FormBuilderComponent.propDecorators = {
    data: [{ type: Input }],
    value: [{ type: Input }],
    id: [{ type: Input }],
    valueChanges: [{ type: Output }],
    validityChanges: [{ type: Output }],
    state: [{ type: Input }]
};
