// @jaspero/ng-helpers/esm2015/directives/click-outside/click-outside.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, NgZone, Output } from '@angular/core';
import { fromEvent } from 'rxjs';
import { filter } from 'rxjs/operators';
/**
 * Emits an event when a click action occurs that does not target the element
 *
 * \@example
 * <div (jpClickOutside)="doSomething()"></div>
 */
export class ClickOutsideDirective {
    /**
     * @param {?} _el
     * @param {?} _ngZone
     */
    constructor(_el, _ngZone) {
        this._el = _el;
        this._ngZone = _ngZone;
        /**
         * Any valid html event
         */
        this.clickOutsideEventType = 'click';
        /**
         * if true jpClickOutside doesn't emit
         */
        this.clickOutsideBlock = false;
        /**
         * Emits when triggered event doesn't contain this e
         */
        this.jpClickOutside = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this._ngZone.runOutsideAngular(() => {
            this.subscription = fromEvent(window, this.clickOutsideEventType)
                .pipe(filter(event => !this.clickOutsideBlock &&
                !this._el.nativeElement.contains(event.target)))
                .subscribe(event => {
                this._ngZone.run(() => {
                    this.jpClickOutside.emit(event);
                });
            });
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
}
ClickOutsideDirective.decorators = [
    { type: Directive, args: [{
                selector: '[jpClickOutside]'
            },] }
];
/** @nocollapse */
ClickOutsideDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgZone }
];
ClickOutsideDirective.propDecorators = {
    clickOutsideEventType: [{ type: Input }],
    clickOutsideBlock: [{ type: Input }],
    jpClickOutside: [{ type: Output }]
};
if (false) {
    /**
     * Any valid html event
     * @type {?}
     */
    ClickOutsideDirective.prototype.clickOutsideEventType;
    /**
     * if true jpClickOutside doesn't emit
     * @type {?}
     */
    ClickOutsideDirective.prototype.clickOutsideBlock;
    /**
     * Emits when triggered event doesn't contain this e
     * @type {?}
     */
    ClickOutsideDirective.prototype.jpClickOutside;
    /** @type {?} */
    ClickOutsideDirective.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    ClickOutsideDirective.prototype._el;
    /**
     * @type {?}
     * @private
     */
    ClickOutsideDirective.prototype._ngZone;
}
