// @jaspero/form-builder/esm2015/lib/fields/draggable-list/draggable-list.component.js
import { moveItemInArray } from '@angular/cdk/drag-drop';
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FieldComponent } from '../../field/field.component';
export class DraggableListComponent extends FieldComponent {
    ngOnInit() {
        if (this.cData.control.value.length) {
            this.cData.options.sort((optionOne, optionTwo) => {
                const indexOne = this.cData.control.value.indexOf(optionOne);
                const indexTwo = this.cData.control.value.indexOf(optionTwo);
                return indexTwo - indexOne;
            });
        }
    }
    drop(event) {
        moveItemInArray(this.cData.options, event.previousIndex, event.currentIndex);
        this.cData.control.setValue(this.cData.options.map(val => {
            return val.value;
        }));
    }
}
DraggableListComponent.decorators = [
    { type: Component, args: [{
                selector: 'fb-draggable-list',
                template: "<div class=\"mat-form-field-wrapper\">\n  <p class=\"fs-small fg-secondary\" *ngIf=\"cData.label\">{{cData.label | transloco}}</p>\n\n  <div\n    class=\"fb-draggable-list\"\n    cdkDropList\n    (cdkDropListDropped)=\"drop($event)\">\n    <div\n      class=\"fb-draggable-list-item\"\n      *ngFor=\"let item of cData.options; index as index\"\n      [cdkDragDisabled]=\"item.disabled\"\n      cdkDrag>\n      {{index + 1}}. {{item.name}}\n    </div>\n  </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".fb-draggable-list{border:1px solid var(--panel-tertiary);border-radius:4px;overflow:hidden}.fb-draggable-list-item{align-items:center;background:var(--panel-theme);border-bottom:1px solid var(--panel-tertiary);cursor:move;display:flex;padding:20px 10px}.fb-draggable-list-item:last-child{border:none}.cdk-drag-preview{border-radius:4px;box-shadow:0 5px 5px -3px rgba(0,0,0,.2),0 8px 10px 1px rgba(0,0,0,.14),0 3px 14px 2px rgba(0,0,0,.12)}.cdk-drag-placeholder{opacity:0}.cdk-drag-animating,.fb-draggable-list.cdk-drop-list-dragging .fb-draggable-list-item:not(.cdk-drag-placeholder){transition:transform .25s cubic-bezier(0,0,.2,1)}"]
            },] }
];
