// @jaspero/ng-helpers/esm2015/directives/track-by-field/track-by-field.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TrackByFieldDirective } from './track-by-field.directive';
export class TrackByFieldModule {
    /**
     * @param {?=} value
     * @return {?}
     */
    static defaultKey(value = 'id') {
        return {
            ngModule: TrackByFieldModule,
            providers: [
                {
                    provide: 'defaultKey',
                    useValue: value
                }
            ]
        };
    }
}
TrackByFieldModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    TrackByFieldDirective
                ],
                exports: [
                    TrackByFieldDirective
                ],
            },] }
];
