// @jaspero/ng-helpers/esm2015/directives/debounce-change/debounce-change.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DebounceChangeDirective } from './debounce-change.directive';
import { DEBOUNCE_TIME } from './debounce-time.const';
export class DebounceChangeModule {
    /**
     * @param {?} value
     * @return {?}
     */
    static defaultDebounceTime(value) {
        return {
            ngModule: DebounceChangeModule,
            providers: [
                {
                    provide: DEBOUNCE_TIME,
                    useValue: value
                }
            ]
        };
    }
}
DebounceChangeModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    DebounceChangeDirective
                ],
                exports: [
                    DebounceChangeDirective
                ],
                providers: [
                    {
                        provide: DEBOUNCE_TIME,
                        useValue: 500
                    }
                ]
            },] }
];
