// @jaspero/ng-helpers/esm2015/directives/debounce-change/debounce-change.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Inject, Input, NgZone, Output } from '@angular/core';
import { fromEvent } from 'rxjs';
import { debounceTime, filter } from 'rxjs/operators';
import { DEBOUNCE_TIME } from './debounce-time.const';
export class DebounceChangeDirective {
    /**
     * @param {?} _el
     * @param {?} _ngZone
     * @param {?} _defaultDebounceTime
     */
    constructor(_el, _ngZone, _defaultDebounceTime) {
        this._el = _el;
        this._ngZone = _ngZone;
        this._defaultDebounceTime = _defaultDebounceTime;
        /**
         * Any valid html event
         */
        this.debounceChangeEventType = 'keyup';
        /**
         * If true and the event has a 'target.value'
         * we listen for it and only emit if the value changed
         */
        this.emitOnlyOnChange = false;
        /**
         * Emits original event after debounce
         */
        this.jpDebounceChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this._ngZone.runOutsideAngular(() => {
            /** @type {?} */
            let prev = this._el.nativeElement.value;
            this.subscription = fromEvent(this._el.nativeElement, this.debounceChangeEventType)
                .pipe(debounceTime(this.debounceTime || this._defaultDebounceTime), filter(event => {
                return event.target &&
                    event.target.value !== undefined &&
                    this.emitOnlyOnChange ?
                    event.target.value !== prev :
                    true;
            }))
                .subscribe(event => {
                this._ngZone.run(() => {
                    if (event.target) {
                        prev = event.target.value;
                    }
                    this.jpDebounceChange.emit(event.target.value);
                });
            });
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
}
DebounceChangeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[jpDebounceChange]'
            },] }
];
/** @nocollapse */
DebounceChangeDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgZone },
    { type: Number, decorators: [{ type: Inject, args: [DEBOUNCE_TIME,] }] }
];
DebounceChangeDirective.propDecorators = {
    debounceTime: [{ type: Input }],
    debounceChangeEventType: [{ type: Input }],
    emitOnlyOnChange: [{ type: Input }],
    jpDebounceChange: [{ type: Output }]
};
if (false) {
    /**
     * time to forward to the debounceTime pipe
     * @type {?}
     */
    DebounceChangeDirective.prototype.debounceTime;
    /**
     * Any valid html event
     * @type {?}
     */
    DebounceChangeDirective.prototype.debounceChangeEventType;
    /**
     * If true and the event has a 'target.value'
     * we listen for it and only emit if the value changed
     * @type {?}
     */
    DebounceChangeDirective.prototype.emitOnlyOnChange;
    /**
     * Emits original event after debounce
     * @type {?}
     */
    DebounceChangeDirective.prototype.jpDebounceChange;
    /** @type {?} */
    DebounceChangeDirective.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    DebounceChangeDirective.prototype._el;
    /**
     * @type {?}
     * @private
     */
    DebounceChangeDirective.prototype._ngZone;
    /**
     * @type {?}
     * @private
     */
    DebounceChangeDirective.prototype._defaultDebounceTime;
}
